gtsam/hybrid/DCMixtureFactor.h

```diff
@@ -34,56 +33,79 @@ namespace gtsam {
  */
 template <class NonlinearFactorType>
 class DCMixtureFactor : public DCFactor {
- private:
-  DiscreteKey dk_;
-  std::vector<NonlinearFactorType> factors_;
-  bool normalized_;
-
  public:
   using Base = DCFactor;
   using shared_ptr = boost::shared_ptr<DCMixtureFactor>;
 
+  using FactorDecisionTree =
+      DecisionTree<Key, boost::shared_ptr<NonlinearFactorType>>;
+
+ private:
+  FactorDecisionTree factors_;
+  bool normalized_;
+
+ public:
   DCMixtureFactor() = default;
 
-  DCMixtureFactor(const KeyVector& keys, const DiscreteKey& dk,
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
+                  const FactorDecisionTree& factors, bool normalized = false)
+      : Base(keys, discreteKeys), factors_(factors), normalized_(normalized) {}
+
+  /**
+   * @brief Convenience constructor that generates the underlying factor
+   * decision tree for us.
+   *
+   * Here it is important that the vector of discrete keys and the vector of
+   * factors have a 1-to-1 mapping so that the decision tree is constructed
+   * accordingly.
+   *
+   * @param keys Vector of keys for continuous factors.
+   * @param discreteKeys Vector of discrete keys.
+   * @param factors Vector of factors. Each factor should map to its
+   * corresponding assigned discrete key.
+   * @param normalized Flag indicating if the factor error is already
+   * normalized.
+   */
+  DCMixtureFactor(const KeyVector& keys, const DiscreteKeys& discreteKeys,
                   const std::vector<NonlinearFactorType>& factors,
                   bool normalized = false)
-      : dk_(dk), factors_(factors), normalized_(normalized) {
-    // Compiler doesn't like `keys_` in the initializer list.
-    keys_ = keys;
-
-    // Add `dk` to `dkeys` list.
-    discreteKeys_.push_back(dk);
+      : Base(keys, discreteKeys), normalized_(normalized) {
+    std::vector<boost::shared_ptr<NonlinearFactorType>> factor_pointers;
```

Comment:
Why was it so much easier for Gaussian mixture factor? I am probably not seeing something. Also, explain this “1-1’ mapping. Look at Signature for an example.

Reply:
Because in `DCGaussianMixtureFactor` we are passing in a vector of shared pointers, so no need to convert to shared pointers like we're doing here.
The 1-to-1 is a misnomer and a mistake. I'll update it.

Reply:
I think this should not be the API. `factors` should be a FactorGraph<NonlinearFactorType>. This conversion costs money, and the rule in GTSAM is: every factor passed around is a shared pointer.

Reply:
Gotcha, but this should probably go in the next PR since the vector API is from the original DC-SAM.